src/integrationTest/java/org/opensearch/security/privileges/actionlevel/SubjectBasedActionPrivilegesTest.java

```diff
@@ -0,0 +1,717 @@
+
+            PrivilegesEvaluatorResponse result = subject.hasExplicitIndexPrivilege(
+                ctx().get(),
+                Set.of("system:admin/system_index"),
+                IndexResolverReplacer.Resolved.ofIndex("test_index")
+            );
+            assertThat(result, isAllowed());
+        }
+
+        @Test
+        public void hasExplicitIndexPrivilege_noWildcard() throws Exception {
+            RoleV7 config = config("""
+                index_permissions:
+                - index_patterns: ['test_index']
+                  allowed_actions: ['*']
+                """);
+            SubjectBasedActionPrivileges subject = new SubjectBasedActionPrivileges(config, FlattenedActionGroups.EMPTY);
+
+            PrivilegesEvaluatorResponse result = subject.hasExplicitIndexPrivilege(
+                ctx().get(),
+                Set.of("system:admin/system_index"),
+                IndexResolverReplacer.Resolved.ofIndex("test_index")
+            );
+            assertThat(result, isForbidden());
+        }
+
+        @Test
+        public void hasExplicitIndexPrivilege_negative_wrongAction() throws Exception {
+            RoleV7 config = config("""
+                index_permissions:
+                - index_patterns: ['test_index']
+                  allowed_actions: ['system:admin/system*']
+                """);
+            SubjectBasedActionPrivileges subject = new SubjectBasedActionPrivileges(config, FlattenedActionGroups.EMPTY);
+
+            PrivilegesEvaluatorResponse result = subject.hasExplicitIndexPrivilege(
+                ctx().get(),
+                Set.of("system:admin/system_foo"),
+                IndexResolverReplacer.Resolved.ofIndex("test_index")
+            );
+            assertThat(result, isForbidden());
+        }
+
+        @Test
+        public void hasExplicitIndexPrivilege_errors() throws Exception {
+            RoleV7 config = config("""
+                index_permissions:
+                - index_patterns: ['/invalid_regex${user.name}\\/']
+                  allowed_actions: ['system:admin/system*']
+                """);
+            SubjectBasedActionPrivileges subject = new SubjectBasedActionPrivileges(config, FlattenedActionGroups.EMPTY);
+
+            PrivilegesEvaluatorResponse result = subject.hasExplicitIndexPrivilege(
+                ctx().get(),
+                Set.of("system:admin/system_index"),
+                IndexResolverReplacer.Resolved.ofIndex("test_index")
+            );
+            assertThat(result, isForbidden());
+            assertTrue(result.hasEvaluationExceptions());
+            assertTrue(
+                "Result contains exception info: " + result.getEvaluationExceptionInfo(),
+                result.getEvaluationExceptionInfo().startsWith("Exceptions encountered during privilege evaluation:")
+            );
+        }
+
+    }
+
+    static RoleV7 config(String config) {
+        return RoleV7.fromYamlStringUnchecked(config);
+    }
+}
```

Comment:
we should test this class with API tokens before that feature goes GA.

Reply:
With "that feature", do you mean API tokens? Or merging this PR?

Reply:
Sure!